src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Encoder.cs

```diff
@@ -493,6 +493,9 @@ private static unsafe void Avx2Encode(ref byte* srcBytes, ref byte* destBytes, b
         [CompExactlyDependsOn(typeof(AdvSimd.Arm64))]
         private static unsafe void AdvSimdEncode(ref byte* srcBytes, ref byte* destBytes, byte* srcEnd, int sourceLength, int destLength, byte* srcStart, byte* destStart)
         {
+#if MONO
```

Comment:
It is better to do this at the callsite of `AdvSimdEncod`.

Reply:
I can add an extra guard to the current caller of AdvSimdEncode. I think the current callsite would just fall through to the next block of logic, so I didn't add it in the first place.

I think we should still have the block on AdvSimdEncode in case more invocations are added, because the PNSE is hit inside of AdvSimdEncode.

Reply:
hhm, wondering why this should be at the callsite rather than just inside `AdvSimdEncode`? Also can you add a comment about why this is under a `#if !MONO` pointing to the https://github.com/dotnet/runtime/issues/96828#issuecomment-1886235203?

In caller, you could just have without `#else`

```c#
#if MONO
 return;
#endif

Reply:
Because at the callsite, there is a check to see if AdvSimd is supported to decide to call `AdvSimdEncode` or not. So I feel logically, it makes more sense to do it there. Functional wise, either way should be fine.

Reply:
> pointing to the [#96828 (comment)](https://github.com/dotnet/runtime/issues/96828#issuecomment-1886235203)?

point it to https://github.com/dotnet/runtime/issues/93081

> there is a check to see if AdvSimd is supported

Right, but that is true even for Mono and should be ok. At least if there are more callers, we won't have to worry about adding the `#if` everywhere.

Reply:
This leads to unreachable code, so adding back in the else

Reply:
this looks good to me, but curious why does `#if MONO` makes rest of the code unreachable for coreclr?

Reply:
I believe the unreachable code was for Mono because it would be the following.
```
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        [CompExactlyDependsOn(typeof(AdvSimd.Arm64))]
        private static unsafe void AdvSimdEncode(ref byte* srcBytes, ref byte* destBytes, byte* srcEnd, int sourceLength, int destLength, byte* srcStart, byte* destStart)
        {
            return; // everything after is unreachable
            <rest of logic>
            ....
            <rest of logic>
         }
```

Reply:
I agree with @fanyang-mono here, this should really be at the callsite of `AdvSimdEncode` and the entire method should likely be inaccessible on Mono. Not only will that help Mono JIT/AOT/LLVM/Interpreter in being more efficient, but it will help make the code easier to follow/understand.

-------------

As an aside, the fact that Mono is reporting `IsSupported` but some functions are not actually supported is also problematic. It means we likely missed running a test or checking in functionality for new APIs somewhere. We should determine where/how that happened so that we can avoid such issues in the future.

Reply:
Updated to guard the entire method

----------

I believe that Mono reporting `IsSupported` but functions not being supported might be attributed to #84510 not accounting for Mono. The volume of mono tests is far too time consuming + non-zero flakey tests to run on all PRs by default. Hence a runtime-extra-platforms lane was introduced, which was intended to be ran for PRs that make changes affecting Mono.

There is a smoke test being ran by default, but that is just System.Runtime.Tests.csproj. We could increase the number of suites that are ran as smoke tests, but that is another question of what other suites should be added that reliably dont flake. It looks like there are HWIntrinsics runtime tests, but guessing it doesn't run for Mono?

Reply:
I don't think mono specific pipelines are run today if the changes are not related to mono. `runtime-extra-platforms` would be a good remainder to run for such cases. Regarding, `IsSupported`, what is a good option here? Although logically correct to have `AdvSimd.IsSupported == false` until those APIs are not implemented in mono, but that would mean that mono won't have any of the other AdvSimd features too.

Reply:
BCL change should trigger mono CI lanes which run library tests to run. If not, it is an issue worth looking into.

Reply:
Since these are new API's, I doubt the customer would use them right away. So I would prefer leaving `AdvSimd.IsSupported == true` as it is now and implement the new API's soon.

Reply:
re: tests, I think this is a case where we didn't have adequate coverage in our smoke tests. The good thing is that we discovered it right away via extra-platforms and we can take action. 

With us running a reduced set of tests for iOS devices (full aot LLVM), this is going to happen. As Mitch was getting at, my suggestion would be to add another suite to the smoke tests that we run so that there's a reasonable chance we would fail on something like this in the future.